src/app/plugins/tictactoe.plugin.ts

```diff
@@ -128,43 +135,110 @@ class TTTGame {
   }
 
   public reset() {
-    this._choosing = Choosing.Row;
+    this._choosing = Choosing.Column;
     this._row = -1;
     this._col = -1;
   }
 
   public async choose(index: number, msg: IMessage) {
-    if (this._choosing === Choosing.Row) {
-      this._row = index;
-      this._choosing = Choosing.Column;
+    if (this._choosing === Choosing.Column) {
+      this._col = index;
+      this._choosing = Choosing.Row;
       await msg.edit(this.showBoard());
       return;
     }
 
-    this._col = index;
+    this._row = index;
 
     // Make the move -------------------
     // Make sure its not overwriting
-    if (this._board[this._col][this._row] !== 0) {
+    if (this._board[this._row][this._col] !== 0) {
       this.reset();
       return;
     }
 
-    this._board[this._col][this._row] = this.currentPlayer;
-    if (this._checkWin()) {
-      this._winner = this.currentPlayer;
-    }
+    this._board[this._row][this._col] = this.currentPlayer;
+    this._checkAndUpdateWin();
 
     this._flipTurn();
     this.reset();
     await msg.edit(this.showBoard());
     // Reset where the player is choosing
+
+    if (!this._gameOver && this.currentPlayer === 1 && this._playingLion) {
+      this._lionMove();
+      this._checkAndUpdateWin();
+
+      this._flipTurn();
+      await msg.edit(this.showBoard());
+    }
+  }
+
+  private _lionMove() {
+    const { bestRow, bestCol } = this._getBestMove();
+    this._board[bestRow][bestCol] = this.currentPlayer;
+  }
+
+  // Get's the strongest move for Lion
+  private _getBestMove() {
+    const moves: { row: number; col: number; val: number }[] = [];
+
+    // For every location
+    this._board.forEach((_, row) =>
+      _.forEach((_, col) => {
+        if (this._board[row][col] !== 0) {
+          return;
+        }
+
+        // Make the move and evaluate the board state
+        this._board[row][col] = this.currentPlayer;
+        moves.push({ row, col, val: this._evaluate(this.currentPlayer * -1) });
+        // Backtrack.
+        this._board[row][col] = 0;
+      })
+    );
+
+    // Sort the moves
+    moves.sort((a, b) => a.val - b.val);
+    return { bestRow: moves[0].row, bestCol: moves[0].col };
+  }
+
+  // Recursive algorithm to find the strength of each possible move
+  private _evaluate(currentPlayer: number) {
+    if (this._checkWin()) {
+      return currentPlayer;
+    }
+
+    if (this._checkTie()) {
+      return 0;
+    }
+
+    const moves: number[] = [];
+    this._board.forEach((_, row) =>
+      _.forEach((_, col) => {
+        if (this._board[row][col] !== 0) {
+          return;
+        }
+        this._board[row][col] = currentPlayer;
+        moves.push(this._evaluate(currentPlayer * -1));
+        this._board[row][col] = 0;
+      })
+    );
```

Comment:
`.reduce`

Reply:
I'll try my best 😂

Reply:
For some reason, using reduce  in this context is modifying the content of `this._board` I don't understand it at all. I also don't see the advantage of reduce in this context, other than removing a single line of code. Why do we prefer reduce over a foreach in this instance?

Reply:
I made the reduce, but it comes out uglier than what is already there